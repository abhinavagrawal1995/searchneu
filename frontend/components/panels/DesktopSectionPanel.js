/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import React from 'react';
import PropTypes from 'prop-types';
import CSSModules from 'react-css-modules';
import classNames from 'classnames/bind';

import globe from './globe.svg';
import css from './DesktopSectionPanel.css';
import Keys from '../../../common/Keys';

import LocationLinks from './LocationLinks';
import WeekdayBoxes from './WeekdayBoxes';

const cx = classNames.bind(css);


class DesktopSectionPanel extends React.Component {
  // Create the 4:35 - 5:40 pm string.
  // This was copied from mobile section panel.js
  // TODO: deduplicate
  getTimeStingFromMeetings(meetingMoments) {
    const times = [];
    meetingMoments.forEach((time) => {
      const startString = time.start.format('h:mm');
      const endString = time.end.format('h:mm a');
      const combinedString = `${startString} - ${endString}`;
      if (!times.includes(combinedString)) {
        times.push(combinedString);
      }
    });
    return times.join(', ');
  }


  render() {
    // Instead of calculating a lot of these individually and putting them together in the return call
    // Append to this array as we go.
    // So the logic can be separated into distinct if statements.
    const tdElements = [];

    // If it is online, just put one super wide cell
    if (this.props.section.online) {
      // How many cells to span
      // need to span more cells if final exam columns are being shown.
      let length = 3;
      if (this.props.shouldShowExamColumns) {
        length = 6;
      }

      const onlineElement =
      (
        <td key='onlineWideCell' colSpan={ length } className={ css.wideOnlineCell }>
          <span className={ css.onlineDivLineContainer }>
            <span className={ `${css.onlineDivLine} ${css.onlineLeftLine}` } />
            <span>Online Class</span>
            <span className={ css.onlineDivLine } />
          </span>
        </td>
      );

      tdElements.push(onlineElement);

    // Have individual cells for the different columns
    } else {
      const meetingMoments = this.props.section.getAllMeetingMoments();
      const meetingStrings = this.getTimeStingFromMeetings(meetingMoments);

      const examMeeting = this.props.section.getExamMeeting();

      let examTimeString = null;
      if (examMeeting) {
        examTimeString = this.getTimeStingFromMeetings(examMeeting.times[0]);
      }


      tdElements.push(<td key='weekDayBoxes'> <WeekdayBoxes section={ this.props.section } /> </td>);
      tdElements.push(<td key='times'>{meetingStrings}</td>);
      tdElements.push(<td key='locationLinks'> <LocationLinks section={ this.props.section } /> </td>);

      // If there are exams, fill in those cells too
      // Calculate the exam elements in each row
      if (this.props.shouldShowExamColumns) {
        const sectionExamMeeting = this.props.section.getExamMeeting();
        if (examMeeting) {
          tdElements.push(<td key='exam1'>{examTimeString}</td>);
          tdElements.push(<td key='exam3'>{sectionExamMeeting.endDate.format('MMM Do')}</td>);
          tdElements.push(<td key='exam4'>{sectionExamMeeting.where}</td>);
        } else {
          tdElements.push(<td key='exam5' />);
          tdElements.push(<td key='exam6' />);
          tdElements.push(<td key='exam7' />);
        }
      }
    }


    return (
      <tr key={ Keys.create(this.props.section).getHash() }>
        <td> {this.props.section.crn} </td>
        <td> {this.props.section.getProfs().join(', ')} </td>

        {tdElements}

        <td>
          <div data-tip='Open Seats/Total Seats' className={ css.inlineBlock }>
            {this.props.section.seatsRemaining}/{this.props.section.seatsCapacity}
          </div>
        </td>

        <td
          className={ cx({
            displayNone: !this.props.showWaitList,
          }) }
        >
          <div data-tip='Open/Total Waitlist Seats' className={ css.inlineBlock }>
            {this.props.section.waitRemaining}/{this.props.section.waitCapacity}
          </div>
        </td>

        <td>
          <a target='_blank' rel='noopener noreferrer' className={ `${css.inlineBlock} ${css.sectionGlobe}` } data-tip={ `View on ${this.props.section.host}` } href={ this.props.section.prettyUrl || this.props.section.url }>
            <img src={ globe } alt='link' />
          </a>
        </td>
      </tr>
    );
  }
}


DesktopSectionPanel.propTypes = {
  shouldShowExamColumns: PropTypes.bool.isRequired,
  showWaitList: PropTypes.bool.isRequired,
  section: PropTypes.object.isRequired,
};


export default CSSModules(DesktopSectionPanel, css);
