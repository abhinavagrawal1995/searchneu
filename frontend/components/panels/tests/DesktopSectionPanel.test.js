/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import React from 'react';

import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import mockData from './mockData';
import DesktopSectionPanel from '../DesktopSectionPanel';

Enzyme.configure({ adapter: new Adapter() });

it('should render a section', () => {
  const result = shallow(<DesktopSectionPanel shouldShowExamColumns={ false } showWaitList={ false } section={ mockData.cs1210.sections[0] } />);
  expect(result).toMatchSnapshot();
});

it('should render another section', () => {
  const result = shallow(<DesktopSectionPanel shouldShowExamColumns={ false } showWaitList={ false } section={ mockData.cs1210.sections[1] } />);
  expect(result).toMatchSnapshot();
});
