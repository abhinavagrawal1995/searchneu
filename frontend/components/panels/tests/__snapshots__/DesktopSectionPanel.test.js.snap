// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render some stuff 1`] = `
<tr>
  <td>
     
    13502
     
  </td>
  <td>
     
    Jennifer Anne Shire
     
  </td>
  <td>
     
    <WeekdayBoxes
      section={
        Section {
          "classId": "1210",
          "classUid": "1210_602555960",
          "crn": "13502",
          "host": "neu.edu",
          "lastUpdateTime": 1511131914029,
          "meetings": Array [
            Meeting {
              "endDate": "2017-12-07T00:00:00.000Z",
              "profs": Array [
                "Jennifer Anne Shire",
              ],
              "startDate": "2017-09-07T00:00:00.000Z",
              "times": Array [
                Array [
                  Object {
                    "end": "1970-01-06T11:30:00.000Z",
                    "start": "1970-01-06T09:50:00.000Z",
                  },
                ],
              ],
              "where": "Snell Engineering Center 108",
            },
          ],
          "online": false,
          "seatsCapacity": 19,
          "seatsRemaining": 0,
          "subject": "CS",
          "termId": "201810",
          "url": "https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_detail_sched?term_in=201810&crn_in=13502",
          "waitCapacity": 0,
          "waitRemaining": 0,
        }
      }
    />
     
  </td>
  <td>
    9:50 - 11:30 am
  </td>
  <td>
     
    <LocationLinks
      section={
        Section {
          "classId": "1210",
          "classUid": "1210_602555960",
          "crn": "13502",
          "host": "neu.edu",
          "lastUpdateTime": 1511131914029,
          "meetings": Array [
            Meeting {
              "endDate": "2017-12-07T00:00:00.000Z",
              "profs": Array [
                "Jennifer Anne Shire",
              ],
              "startDate": "2017-09-07T00:00:00.000Z",
              "times": Array [
                Array [
                  Object {
                    "end": "1970-01-06T11:30:00.000Z",
                    "start": "1970-01-06T09:50:00.000Z",
                  },
                ],
              ],
              "where": "Snell Engineering Center 108",
            },
          ],
          "online": false,
          "seatsCapacity": 19,
          "seatsRemaining": 0,
          "subject": "CS",
          "termId": "201810",
          "url": "https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_detail_sched?term_in=201810&crn_in=13502",
          "waitCapacity": 0,
          "waitRemaining": 0,
        }
      }
    />
     
  </td>
  <td>
    <div
      className="inlineBlock"
      data-tip="Open Seats/Total Seats"
    >
      0
      /
      19
    </div>
  </td>
  <td
    className="displayNone"
  >
    <div
      className="inlineBlock"
      data-tip="Open/Total Waitlist Seats"
    >
      0
      /
      0
    </div>
  </td>
  <td>
    <a
      className="inlineBlock sectionGlobe"
      data-tip="View on neu.edu"
      href="https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_detail_sched?term_in=201810&crn_in=13502"
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        alt="link"
        src="file"
      />
    </a>
  </td>
</tr>
`;
